lib/rmg/CRmgTemplateZone.cpp

```diff
@@ -243,6 +262,481 @@ void CRmgTemplateZone::createBorder()
 	}
 }
 
+void CRmgTemplateZone::createWater(EWaterContent::EWaterContent waterContent, bool debug)
+{
+	if(waterContent == EWaterContent::NONE || isUnderground())
+		return; //do nothing
+	
+	std::set<int3> waterTiles = collectDistantTiles((float)(getSize() + 1));
+	
+	//add border tiles as water for ISLANDS
+	if(waterContent == EWaterContent::ISLANDS)
+	{
+		for (auto& tile : tileinfo)
+		{
+			if (gen->shouldBeBlocked(tile))
+			{
+				waterTiles.insert(tile);
+			}
+		}
+	}
+	
+	std::list<int3> tilesQueue(waterTiles.begin(), waterTiles.end()); //tiles need to be processed
+	std::set<int3> tilesChecked = waterTiles; //tiles already processed
+	std::map<int, std::set<int3>> coastTilesMap; //key: distance to water; value: tiles with that distance
+	std::map<int3, int> tilesDist; //key: tile; value: distance to water
+	
+	//optimization: prefill distance for all tiles marked for water with 0
+	for(auto& tile : waterTiles)
+	{
+		tilesDist[tile] = 0;
+	}
+	
+	//lambda for increasing distance of negihbour tiles
+	auto coastSearch = [this, &tilesDist, &tilesChecked, &coastTilesMap, &tilesQueue](const int3 & src, const int3 & dst)
+	{
+		if(!gen->map->isInTheMap(dst))
+			return;
+		
+		if(tilesChecked.find(dst)!=tilesChecked.end())
+			return;
+		
+		if(tileinfo.find(dst) != tileinfo.end())
+		{
+			tilesDist[dst] = tilesDist[src] + 1;
+			coastTilesMap[tilesDist[dst]].insert(dst);
+			tilesChecked.insert(dst);
+			tilesQueue.push_back(dst);
+		}
+	};
+	
+	//fills the distance-to-water map
+	while(!tilesQueue.empty())
+	{
+		int3 tile = tilesQueue.front();
+		tilesQueue.pop_front();
+		gen->foreachDirectNeighbour(tile, std::bind(coastSearch, tile, std::placeholders::_1));
```

Comment:
Lambdas are great, but why do you ever use std::placeholders? :O

Just define anonymous lambda in-place, like here for instance:
```
		gen->foreachDirectNeighbour(tile, [this, &waterTiles, &groundCoast](const int3 & t)
		{
			if(waterTiles.find(t) == waterTiles.end() && tileinfo.find(t) != tileinfo.end()) //for ground tiles of same zone
			{
				groundCoast.push_back(t);
			}
		});
```

Reply:
It's sort of optimization: no need to recreate lambda every iteration in cycle

Reply:
Anonymous fucntion is created once and passed via value to inner loop, where it's called directly.

Reply:
> Lambdas are great, but why do you ever use std::placeholders? :O
> 
> Just define anonymous lambda in-place, like here for instance:
> 
> ```
> gen->foreachDirectNeighbour(tile, [this, &waterTiles, &groundCoast](const int3 & t)
> {
> 	if(waterTiles.find(t) == waterTiles.end() && tileinfo.find(t) != tileinfo.end()) //for ground tiles of same zone
> 	{
> 		groundCoast.push_back(t);
> 	}
> });
> ```

They are but I do not feel like lambda is a separate method because it can be just everywhere inside of a class method making the method body very big. If at least all lambdas are on the top of makeWater - it is more or less ok. But they are in the middle mixed with some other pieces of code. Also presence of comments show clearly that the method is way too big.

Reply:
I don't think placeholders are needed anymore.
If you don't wanna a big inline lambda, just replace `std::bind(coastSearch, tile, std::placeholders::_1)` with `[&tile](const auto& t) { coastSearch(tile, t); }` here, which is much easier to read.

Reply:
I am not against big inline lambdas but against big methods and I treat lambda as a part of its hosting method. So I also have such big methods. It is so ease to start developing something using inline lambdas and classes, everything in a single file. Unfortunately Visual Studio does not have a button like "Move all types to separate files" so refactoring such things takes some time.